ably/realtime_channel.go

```diff
@@ -2,130 +2,342 @@ package ably
+}
+
+func newChannels(client *RealtimeClient) *Channels {
+	return &Channels{
+		client: client,
+		chans:  make(map[string]*RealtimeChannel),
+	}
+}
+
+// Get looks up a channel given by the name and creates it if it does not exist
+// already.
+//
+// It is safe to call Get from multiple goroutines - a single channel is
+// guaranteed to be created only once for multiple calls to Get from different
+// goroutines.
+func (ch *Channels) Get(name string) *RealtimeChannel {
+	ch.mtx.Lock()
+	c, ok := ch.chans[name]
+	if !ok {
+		c = newRealtimeChannel(name, ch.client)
+		ch.chans[name] = c
 	}
+	ch.mtx.Unlock()
+	return c
 }
 
-type ChanState int
+// All returns a list of created channels.
+//
+// It is safe to call All from multiple goroutines, however there's no guarantee
+// the returned list would not list a channel that was already released from
+// different goroutine.
+//
+// The returned list is sorted by channel names.
+func (ch *Channels) All() []*RealtimeChannel {
+	ch.mtx.Lock()
+	chans := make([]*RealtimeChannel, 0, len(ch.chans))
+	for _, c := range ch.chans {
+		chans = append(chans, c)
+	}
+	ch.mtx.Unlock()
+	chanSlice(chans).Sort()
+	return chans
 
-const (
-	ChanStateInitialized ChanState = iota
-	ChanStateAttaching
-	ChanStateAttached
-	ChanStateDetaching
-	ChanStateDetached
-	ChanStateFailed
-)
+}
+
+// Release closes a channel looked up by the name.
+//
+// It is safe to call Release from multiple goroutines - if a channel happened
+// to be already concurrently released, the method is a nop.
+func (ch *Channels) Release(name string) error {
+	ch.mtx.Lock()
+	defer ch.mtx.Unlock()
+	if c, ok := ch.chans[name]; ok {
+		delete(ch.chans, name)
+		return c.Close()
+	}
+	return nil
+}
 
+// RealtimeChannel represents a single named message channel.
+//
+// The name of the channel is read-only - assigning new name to it does not
+// have any effect.
 type RealtimeChannel struct {
-	Name string
+	Name string // read-only name of the channel
 
-	client *RealtimeClient
+	client    *RealtimeClient
+	state     *stateEmitter
+	err       error
+	listen    map[string]map[chan *proto.Message]struct{}
+	listenMtx sync.Mutex
+	queue     *msgQueue
+}
+
+func newRealtimeChannel(name string, client *RealtimeClient) *RealtimeChannel {
+	c := &RealtimeChannel{
+		Name:   name,
+		client: client,
+		state:  newStateEmitter(StateChan, StateChanInitialized, name),
+		listen: make(map[string]map[chan *proto.Message]struct{}),
+	}
+	c.queue = newMsgQueue(client.Connection)
+	if c.client.opts.Listener != nil {
+		c.On(c.client.opts.Listener)
+	}
+	return c
+}
+
+var errAttach = newError(90000, errors.New("ably: Attach on inactive connection"))
 
-	State    ChanState
-	stateMtx sync.Mutex
+// Attach initiates attach request, which is being processed on a separate
+// goroutine.
+//
+// In order to receive its result register a channel with the On method; upon
+// successful attachment the channel will receive StatChanAttached event.
```

Comment:
Done.

Upon Attach return, the attach message is sent and the state is either Attaching or Failed. Moreover to make it more user-friendly, it returns a Result value (in Java library it’s called CompletionListener) which can be used to wait until channel becomes attached or when it transits to a failed / closed state.

(I had no better idea for better name than Result, if you see any please suggest and I’ll change)